README.md
# Memoize

[![CircleCI](https://circleci.com/gh/5app/memoize.svg?style=shield)](https://circleci.com/gh/5app/memoize)

> In computing, memoization or memoisation is an optimization technique used primarily to speed up computer programs by storing the results of expensive function calls and returning the cached result when the same inputs occur again. Memoization has also been used in other contexts (and for purposes other than speed gains), such as in simple mutually recursive descent parsing.[1] Although related to caching, memoization refers to a specific case of this optimization, distinguishing it from forms of caching such as buffering or page replacement. In the context of some logic programming languages, memoization is also known as tabling.[2]

<cite>https://en.wikipedia.org/wiki/Memoization</cite>

# Usage

!Warning contrived example ahead...

```js
import got from 'got'; // simple http requst library for the purpose of demonstration
import memoize from '@5app/memoize';

// Let's decorate the got function
const memoGot = memoize(got);

// Simultaneously open two connections...
const link = 'https://github.com';
Promise.all([memoGot(link), memoGot(link)];

// Will call...
// GET https://github.com
// ... but that's it, it wont call it again the second request will piggy back off the first.
```

# Options `memoize(handler, {...options})`

-   `option.useCache` _(Boolean|Function)_: A truthy/fasly or a function to decide whether to use the cached record or not. Default `true`
-   `option.staleInMs` _Number_: The number of milliseconds before the cache is deemed stale. Results will still be served from the cache whilst an attempt to refresh the cache is made separatly. Default `10000` ms.
-   `option.getKey` _Function_: A function to create a key based upon the input of the function being memoized. Default: a serialization of all the arguments.
-   `option.cache` _Object_: Instance of a Map like object to store the cache. Default `new Map`
-   `options.cacheMaxSize` _Number_: The maximum number of entries to store in the cache. Default `1000`

## `option.useCache`

Whether to use cache this can be a Boolean value (useful to disable it when testing). Or a function e.g.

This snippet checks the cached value before deciding whether to use it...

```js
import memoize from '@5app/memoize';

const memoGot = memoize(got, {
	/**
 	 * @param {object} cached_response - Cached Object
	 * @param {number} cached_response.timestamp - Timestamp when request resolved
	 * @param {string} cached_response.status - 'pending', 'fullfilled', 'rejected'
	 * @param {Promise<*>} cached_response.value - Promise of the request
	 * @returns {Boolean}
	 */
	useCache({timestamp, status}) {
		// Set an expiry on the cache.
		// 2xx, 3xx response last for a full minute before being reused
		// 4xx, 5xx last only a second...
		const age = value.statusCode >= 400 ? 1000 : 60000;

		// Return true if the cache is un-expired, else false.
		return timestamp > Date.now() - AGE;
	}
}

// ...

// Use Memogot
// const req = memogot('link');
// ...

```

# Globally Disable

In testing to disable the memoize cache which can cause issues, use the `MEMOIZE_DISABLE` environment variable, e.g.

```bash
MEMOIZE_DISABLE=1
```

When set to a truthy value the memoize cache will be circumvented.


CHANGELOG.md
# [2.0.0](https://github.com/5app/memoize/compare/v1.4.7...v2.0.0) (2023-03-17)


### Features

* **typescript:** migrate to typescript ([#56](https://github.com/5app/memoize/issues/56)) ([52fa531](https://github.com/5app/memoize/commit/52fa5315f017ac3b07736cef6b44bd326dca28b3))


### BREAKING CHANGES

* **typescript:** this changes the getKey parameters 😭

* chore(deps): update deps

* feat(ts): add callback values, noissue

* chore: rm editorconfig

* chore(lint): fix lint

* chore(ts): remove deprecation warnings


BREAKING CHANGES: Change in the `options.getKey(...args)` => `options.getKeys(args)` args is contained in first parameter.

## [1.4.7](https://github.com/5app/memoize/compare/v1.4.6...v1.4.7) (2022-09-07)


### Bug Fixes

* **github-packages:** test release/npm_and_github circleci orb ([f081604](https://github.com/5app/memoize/commit/f081604c154671c76fc9a38d6e0bbf68ede78a15))

## [1.4.6](https://github.com/5app/memoize/compare/v1.4.5...v1.4.6) (2022-09-06)


### Bug Fixes

* **github-packages:** test release/npm_and_github circleci orb ([97bbdef](https://github.com/5app/memoize/commit/97bbdef0074addf988a76570985a29b342bb548a))

## [1.4.5](https://github.com/5app/memoize/compare/v1.4.4...v1.4.5) (2022-09-05)


### Bug Fixes

* **github-packages:** publish github registry ([09cb723](https://github.com/5app/memoize/commit/09cb72307c933b71c4dfe9b24d86b15389a76f20))

## [1.4.4](https://github.com/5app/memoize/compare/v1.4.3...v1.4.4) (2022-09-05)


### Bug Fixes

* **github-packages:** publish github registry experiment ([3fc4152](https://github.com/5app/memoize/commit/3fc4152b881ace22b1638d8a7f33c2fd42cb1892))

## [1.4.3](https://github.com/5app/memoize/compare/v1.4.2...v1.4.3) (2022-09-05)


### Bug Fixes

* **github-packages:** publish github registry ([e9bef64](https://github.com/5app/memoize/commit/e9bef649241ac59c88ae30d55351b566f07bed6e))

## [1.4.2](https://github.com/5app/memoize/compare/v1.4.1...v1.4.2) (2022-09-05)


### Bug Fixes

* **github-packages:** publish github registry ([9566065](https://github.com/5app/memoize/commit/956606527f38c42889b75c3fb1e027f0195f1d6c))

## [1.4.1](https://github.com/5app/memoize/compare/v1.4.0...v1.4.1) (2022-09-05)


### Bug Fixes

* **github-packages:** publish github registry ([15b7fbb](https://github.com/5app/memoize/commit/15b7fbb20fd88eac4263e0c4bbd5675a687e2c3d))

# [1.4.0](https://github.com/5app/memoize/compare/v1.3.1...v1.4.0) (2022-08-03)


### Features

* **esm:** convert to esm, npm update, noissue ([#40](https://github.com/5app/memoize/issues/40)) ([e391722](https://github.com/5app/memoize/commit/e391722a678694d49d3401fa19b8ead7f08a3b51))

## [1.3.1](https://github.com/5app/memoize/compare/v1.3.0...v1.3.1) (2021-02-17)


### Bug Fixes

* MEMOIZE_DISABLED - rm 'D' from env var ([d02ac0b](https://github.com/5app/memoize/commit/d02ac0b9cf9a11b38e522876883169b4020948e6))

# [1.3.0](https://github.com/5app/memoize/compare/v1.2.0...v1.3.0) (2021-02-16)


### Features

* **cache:** Disable Env Var MEMOIZE_DISABLE ([408170e](https://github.com/5app/memoize/commit/408170e8cbb9a4e888652b5453d825f4f25c0a5b))

# [1.2.0](https://github.com/5app/memoize/compare/v1.1.0...v1.2.0) (2020-07-30)


### Features

* **cacheMaxSize:** Remove oldest items from cache ([a92d21d](https://github.com/5app/memoize/commit/a92d21d260e5654a40c609a94a9a79424cd4a838))

# [1.1.0](https://github.com/5app/memoize/compare/v1.0.2...v1.1.0) (2020-02-21)


### Features

* **useCached:** update useCached to accept a function or boolean ([bfbdd8f](https://github.com/5app/memoize/commit/bfbdd8f77c8e49839f65da005733e5b2aba08714))

## [1.0.2](https://github.com/5app/memoize/compare/v1.0.1...v1.0.2) (2020-02-21)


### Bug Fixes

* **deps:** move dev dependency ([4741332](https://github.com/5app/memoize/commit/4741332e32d7c59e61247fae7ff4fb4a5fb5021d))

## [1.0.1](https://github.com/5app/memoize/compare/v1.0.0...v1.0.1) (2020-02-21)


### Bug Fixes

* **jsdoc:** option descriptions ([3ac9481](https://github.com/5app/memoize/commit/3ac9481627e147d2c308257524c26f9d9bfdb78b))

# 1.0.0 (2020-02-21)


### Features

* **init:** initial release ([a736337](https://github.com/5app/memoize/commit/a736337c97bee99fe4a254cdbdba16a6fe66bd8e))

## [1.2.1](https://github.com/5app/js-template/compare/v1.2.0...v1.2.1) (2019-12-12)


### Bug Fixes

* **deps:** npm audit ([ca738db](https://github.com/5app/js-template/commit/ca738dba9044e54931fab71afdee889a8acde958))

# [1.2.0](https://github.com/5app/js-template/compare/v1.1.1...v1.2.0) (2019-12-12)


### Features

* **prettier:** format file commits using prettier ([#22](https://github.com/5app/js-template/issues/22)) ([7f6eee8](https://github.com/5app/js-template/commit/7f6eee8f884fa4b21a7799df4b6727ab0a430415))

## [1.1.1](https://github.com/5app/js-template/compare/v1.1.0...v1.1.1) (2019-11-06)


### Bug Fixes

* **publish:** add scope for npm package ([4235eed](https://github.com/5app/js-template/commit/4235eed4a50b3ef4b7e8d1c949b25c1a8ce3ad11))

# [1.1.0](https://github.com/5app/js-template/compare/v1.0.0...v1.1.0) (2019-11-06)


### Features

* **publish:** set publishConfig access public ([2584e1f](https://github.com/5app/js-template/commit/2584e1f3bc90827ecf98ec7ca7286facdf1d9baf))

# 1.0.0 (2019-06-27)


### Features

* **release:** Add semantic release to the mix ([1f6caa6](https://github.com/5app/js-template/commit/1f6caa6))


test/spec/memoize.js
import chai from 'chai';
import chaiAsPromised from 'chai-as-promised';
import memoize from '../../dist/index.js';

// Extend Chai
chai.use(chaiAsPromised);

const {expect} = chai;

describe('memoize', () => {
	beforeEach(() => {
		delete process.env.MEMOIZE_DISABLE;
	});

	it('should decorate a function and return a function', () => {
		const mem = memoize(() => {
			// Continue
		});
		expect(mem).to.be.a('function');
	});
	it('should prevent multiple calls to a function', async () => {
		let counter = 0;
		const mem = memoize(async () => {
			counter++;
		});

		await mem(1);
		await mem(1);

		expect(counter).to.equal(1);
	});

	it('should ignore cache when option.useCached = false', async () => {
		let counter = 0;
		const mem = memoize(
			async () => {
				counter++;
			},
			{
				useCached: false,
			}
		);

		// When synchronous, both are executed..
		await mem(1);
		await mem(1);
		expect(counter).to.equal(2);

		// But when their asynchronous requests should only bump by 1
		counter = 0;
		await Promise.all([mem(1), mem(1)]);
		expect(counter).to.equal(1);
	});

	it('should trigger a rerun when request is stale', async () => {
		let counter = 0;
		const mem = memoize(
			async () => {
				counter++;
				return counter;
			},
			{
				staleInMs: 0,
			}
		);

		// When synchronous, both are executed..
		const valueA = await mem(1);
		const valueB = await mem(1);

		// The response are served from the first request
		expect(valueA).to.equal(1);
		expect(valueB).to.equal(1);

		// But the second is run pending a response
		expect(counter).to.equal(2);
	});

	it('should trigger a rerun when request is stale', async () => {
		let counter = 0;
		const mem = memoize(
			async () => {
				counter++;
				return counter;
			},
			{
				staleInMs: 0,
			}
		);

		// Each request is marked as stale
		// Returns 1 as it's the first request
		const valueA = await mem(1);
		expect(valueA).to.equal(1);

		// Also returns 1, rather than wait for it's own response
		const valueB = await mem(1);
		expect(valueB).to.equal(1);

		// Returns 2, because the second run has now finished.
		const valueC = await mem(1);
		expect(valueC).to.equal(2);
	});

	it('should skip rejected requests', async () => {
		let counter = 0;
		const mem = memoize(
			async () => {
				counter++;
				if (counter % 2) {
					throw new Error('this is odd');
				}
				return counter;
			},
			{
				staleInMs: 0,
			}
		);

		// 1. Fails
		await expect(mem(1)).to.be.eventually.rejectedWith('this is odd');

		// 2. Succeeds returns 2
		const valueB = await mem(1);
		expect(valueB).to.equal(2);

		// 3. Rejects but returns 2
		const valueC = await mem(1);
		expect(valueC).to.equal(2);
	});

	it('should skip rejected responses when useCache:true', async () => {
		let counter = 0;
		const mem = memoize(
			async () => {
				counter++;
				if (counter % 2) {
					throw new Error('this is odd');
				}
				return counter;
			},
			{
				useCache: true,
				staleInMs: 0,
			}
		);

		// 1. Fails
		await expect(mem(1)).to.be.eventually.rejectedWith('this is odd');

		// 2. Succeeds returns 2
		const valueB = await mem(1);
		expect(valueB).to.equal(2);

		// 3. Rejects but returns 2
		const valueC = await mem(1);
		expect(valueC).to.equal(2);
	});

	it('should allow control over whether to use the Cache', async () => {
		let counter = 0;
		const mem = memoize(
			async () => {
				counter++;
				throw new Error(`this is odd ${counter}`);
			},
			{
				useCached() {
					// Use cache of rejected as well
					return true;
				},
				staleInMs: 10_000, // sometime in the distance
			}
		);

		// 1. Fails
		await expect(mem(1)).to.be.eventually.rejectedWith('this is odd 1');

		// 2. Fails, but doesn't try again
		await expect(mem(1)).to.be.eventually.rejectedWith('this is odd 1');
	});

	it('should not contain more than 1000 enties in the cache by default', async () => {
		const cache = new Map();
		const mem = memoize(async number => number / 2, {
			cache,
		});

		await Promise.all(
			Array(2000)
				.fill(null)
				.map((_, index) => mem(index))
		);

		expect(cache.size).to.equal(1000);
	});

	it('should no longer cache when process.env.MEMOIZE_DISABLE=true', async () => {
		let i = 0;

		process.env.MEMOIZE_DISABLE = true;

		const mem = memoize(async () => i++);

		const a = await mem(1);
		const b = await mem(1);

		expect(a).to.equal(0);
		expect(b).to.equal(1);
	});
});


index.js
/**
 * Memoize
 * Memoization or memoisation is an optimization technique used primarily to speed up
 * computer programs by storing the results of expensive function calls and returning
 * the cached result when the same inputs occur again.
 *
 * This function acts as a javascript decorator around functions so they can be used as documented
 *
 * It handles updates (reruns) when
 * - The previous result is regarded as stale
 * - The previous result returned a rejected promise
 *
 *
 * @param {Function} callback - Function to wrap
 * @param {object} opts - Options
 * @param {boolean|Function} [opts.useCached=true] - Whether to serve from previous resolved cached responses
 * @param {number} [opts.staleInMs=10000] - Number of milliseconds before the item is regarded as stale
 * @param {Function} [opts.getKey=JSON.stringify] - Function used to define the key for use
 * @param {object} [opts.cache=new Map()] - Caching function uses Map by default
 * @param {number} [opts.cacheMaxSize=1000] - Maximum Cache Size
 * @returns {Function} The decorated callback function
 */
function Memoize(callback, opts = {}) {
	// Disable all memoize
	const {MEMOIZE_DISABLE = false} = process.env;

	const {
		// useCached: Will return the last resolved cached value
		useCached = true,

		// staleInMs: How long before we should check for new updates
		staleInMs = 10000,

		// getKey: Default key definition
		getKey = JSON.stringify,

		// cache: Caching Map
		cache = new Map(),

		// cache Max Size
		cacheMaxSize = 1000,
	} = opts;

	// Default check for shouldUseCache
	// If we have a resolved value, but we want to keep it up to date set to true
	let shouldUseCache = item => item.status === 'resolved' && useCached;

	// If the settings say it's a function use that instead
	if (typeof useCached === 'function') {
		// Update the condition to use the bespoke option
		shouldUseCache = useCached;
	}

	/**
	 * Decorator
	 *
	 * @param {...*} args - Any number of arguments
	 * @returns {Promise} Value of the callback
	 */
	return async (...args) => {
		// Serialize the keys...
		const key = getKey(...args);

		// Find value based upon the key
		const item = cache.get(key) || {};

		// Has value resolved yet?
		if (item.status === 'pending') {
			// Return the promise
			return item.value;
		}

		// Serve the cached value?
		if (item.value && !MEMOIZE_DISABLE && shouldUseCache(item)) {
			// We're going to return the last resolved value
			// But first let's identify the next request
			// And ensure it's been long enough since the last one...
			if (!item.next && item.timestamp <= Date.now() - staleInMs) {
				const next = callback(...args);
				item.next = next;
				next.then(
					() => {
						// Overwrite the value based upon a positive response
						item.value = next;
						item.timestamp = Date.now();
					},
					() => ({
						// Do nothing if it's unsuccessful
					})
				).finally(() => {
					// Finally cleanup the next marker
					delete item.next;
				});
			}

			// Return the resolved response
			return item.value;
		}

		// Else call the handler
		const value = callback(...args);

		// Update the item with the request
		item.value = value;

		// Set a timestamp
		item.timestamp = Date.now();

		// Set default status
		item.status = 'pending';

		// Update it's status when it, resolves/rejects
		value.then(
			() => {
				item.status = 'resolved';
			},
			() => {
				item.status = 'rejected';
			}
		);

		// Store the promise for use later...
		cache.set(key, item);

		// Does the cache need a trim?
		if (cacheMaxSize && cache.size > cacheMaxSize) {
			for (const k of cache.keys()) {
				// Remove the first key and break
				cache.delete(k);
				break;
			}
		}

		// Return the item value
		return item.value;
	};
}

module.exports = Memoize;


src/index.ts
/* eslint-disable @typescript-eslint/no-explicit-any */
interface MemoizeCacheItem<CallbackValue> {
	// Memoize item status
	status?: 'pending' | 'resolved' | 'rejected';

	// Value if resolved
	value?: CallbackValue;

	// Next request in waiting
	next?: CallbackValue;

	// When this cache item was created
	timestamp?: number;
}

type MemoizeCacheKey = string | number | symbol;

type MemoizeUseCacheHandler<CallbackValue> = (
	param: MemoizeCacheItem<CallbackValue>
) => boolean;

interface MemoizeOptions<ParameterTypes, CallbackValue> {
	// useCached: Will return the last resolved cached value
	useCached?: true | false | MemoizeUseCacheHandler<CallbackValue>;

	// staleInMs: How long before we should check for new updates
	staleInMs?: number;

	// getKey: Default key definition
	getKey?: (args: ParameterTypes) => MemoizeCacheKey;

	// cache: Caching Map
	cache?: Map<MemoizeCacheKey, MemoizeCacheItem<CallbackValue>>;

	// cache Max Size
	cacheMaxSize?: number;
}

/**
 * Memoize
 * Memoization or memoisation is an optimization technique used primarily to speed up
 * computer programs by storing the results of expensive function calls and returning
 * the cached result when the same inputs occur again.
 *
 * This function acts as a javascript decorator around functions so they can be used as documented
 *
 * It handles updates (reruns) when
 * - The previous result is regarded as stale
 * - The previous result returned a rejected promise
 *
 *
 * @param {Function} callback - Function to wrap
 * @param {object} opts - Options
 * @param {boolean|Function} [opts.useCached=true] - Whether to serve from previous resolved cached responses
 * @param {number} [opts.staleInMs=10000] - Number of milliseconds before the item is regarded as stale
 * @param {Function} [opts.getKey=JSON.stringify] - Function used to define the key for use
 * @param {object} [opts.cache=new Map()] - Caching function uses Map by default
 * @param {number} [opts.cacheMaxSize=1000] - Maximum Cache Size
 * @returns {Function} The decorated callback function
 */
export default function Memoize<
	ParameterTypes extends Array<unknown>,
	CallbackValue extends Promise<unknown>
>(
	callback: (...args: ParameterTypes) => CallbackValue,
	opts: MemoizeOptions<ParameterTypes, CallbackValue> = {}
) {
	// Disable all memoize
	const {MEMOIZE_DISABLE = false} = process.env;

	const {
		// useCached: Will return the last resolved cached value
		useCached = true,

		// staleInMs: How long before we should check for new updates
		staleInMs = 10_000,

		// getKey: Default key definition
		getKey = JSON.stringify,

		// cache: Caching Map
		cache = new Map(),

		// cache Max Size
		cacheMaxSize = 1000,
	}: MemoizeOptions<ParameterTypes, CallbackValue> = opts;

	// Default check for shouldUseCache
	// If we have a resolved value, but we want to keep it up to date set to true
	let shouldUseCache: MemoizeUseCacheHandler<CallbackValue>;

	// If the settings say it's a function use that instead
	if (typeof useCached === 'function') {
		// Update the condition to use the bespoke option
		shouldUseCache = useCached;
	} else {
		shouldUseCache = item => item.status === 'resolved' && useCached;
	}

	/**
	 * Decorator
	 *
	 * @param {...*} args - Any number of arguments
	 * @returns {Promise} Value of the callback
	 */
	return async (...args: ParameterTypes) => {
		// Serialize the keys...
		const key = getKey(args);

		// Find value based upon the key
		const item: MemoizeCacheItem<CallbackValue> = cache.get(key) || {};

		// Has value resolved yet?
		if (item.status === 'pending') {
			// Return the promise
			return item.value;
		}

		// Serve the cached value?
		if (item.value && !MEMOIZE_DISABLE && shouldUseCache(item)) {
			// We're going to return the last resolved value
			// But first let's identify the next request
			// And ensure it's been long enough since the last one...
			if (
				!item.next &&
				item.timestamp &&
				item.timestamp <= Date.now() - staleInMs
			) {
				const next = callback(...args);
				item.next = next;
				next.then(
					() => {
						// Overwrite the value based upon a positive response
						item.value = next;
						item.timestamp = Date.now();
					},
					() => ({
						// Do nothing if it's unsuccessful
					})
				).finally(() => {
					// Finally cleanup the next marker
					delete item.next;
				});
			}

			// Return the resolved response
			return item.value;
		}

		// Else call the handler
		const value = callback(...args);

		// Update the item with the request
		item.value = value;

		// Set a timestamp
		item.timestamp = Date.now();

		// Set default status
		item.status = 'pending';

		// Update it's status when it, resolves/rejects
		value.then(
			() => {
				item.status = 'resolved';
			},
			() => {
				item.status = 'rejected';
			}
		);

		// Store the promise for use later...
		cache.set(key, item);

		// Does the cache need a trim?
		if (cacheMaxSize && cache.size > cacheMaxSize) {
			for (const k of cache.keys()) {
				// Remove the first key and break
				cache.delete(k);
				break;
			}
		}

		// Return the item value
		return item.value;
	};
}
